packages/@aws-cdk/aws-codepipeline-actions/lib/cloudformation/stackset-action.ts

```diff
@@ -0,0 +1,691 @@
+ * test your deployment and view the generated stack set, then add the
+ * CloudFormationStackInstances action so that the remaining instances are
+ * created and updated.
+ */
+export class CloudFormationStackSetAction extends Action {
+  private readonly props: CloudFormationStackSetActionProps;
+  private readonly deploymentModel: StackSetDeploymentModel;
+
+  constructor(props: CloudFormationStackSetActionProps) {
+    super({
+      ...props,
+      provider: 'CloudFormationStackSet',
+      category: codepipeline.ActionCategory.DEPLOY,
+      artifactBounds: {
+        minInputs: 1,
+        maxInputs: 3,
+        minOutputs: 0,
+        maxOutputs: 0,
+      },
+      inputs: [
+        props.templatePath.artifact,
+        ...props.parameters?.artifactsReferenced ?? [],
+        ...props.stackInstances?.artifactsReferenced ?? [],
+      ],
+    });
+
+    this.props = props;
+    this.deploymentModel = props.deploymentModel ?? StackSetDeploymentModel.selfManaged();
+  }
+
+  protected bound(scope: CoreConstruct, _stage: codepipeline.IStage, options: codepipeline.ActionBindOptions): codepipeline.ActionConfig {
+    const singletonPolicy = SingletonPolicy.forRole(options.role);
+    singletonPolicy.grantCreateUpdateStackSet(this.props);
+
+    const instancesResult = this.props.stackInstances?.bind(scope);
+    const permissionModelBind = this.deploymentModel?.bind(scope);
+
+    for (const role of permissionModelBind?.passedRoles ?? []) {
+      singletonPolicy.grantPassRole(role);
+    }
+
+    if ((this.actionProperties.inputs || []).length > 0) {
+      options.bucket.grantRead(singletonPolicy);
+    }
+
+    return {
+      configuration: {
+        StackSetName: this.props.stackSetName,
+        Description: this.props.description,
+        TemplatePath: this.props.templatePath.location,
+        Parameters: this.props.parameters?.render(),
+        Capabilities: parseCapabilities(this.props.cfnCapabilities),
+        FailureTolerancePercentage: this.props.failureTolerancePercentage,
+        MaxConcurrentPercentage: this.props.maxConcurrentPercentage,
+        ...instancesResult?.stackSetConfiguration,
+        ...permissionModelBind?.stackSetConfiguration,
+      },
+    };
+  }
+}
+
+/**
+ * Properties for the CloudFormationStackInstancesAction
+ */
+export interface CloudFormationStackInstancesActionProps extends codepipeline.CommonAwsActionProps {
+  /**
+   * The AWS region the given Action resides in.
+   * Note that a cross-region Pipeline requires replication buckets to function correctly.
+   * You can provide their names with the {@link PipelineProps#crossRegionReplicationBuckets} property.
+   * If you don't, the CodePipeline Construct will create new Stacks in your CDK app containing those buckets,
+   * that you will need to `cdk deploy` before deploying the main, Pipeline-containing Stack.
+   *
+   * @default - the Action resides in the same region as the Pipeline
```

Comment:
I have to say, I don't quite understand how this works...? Is this the `region` of the StackSet itself? I don't think it's the region of the StackSet instances, right? Because those are set in `stackInstances` below...

If my hunch is correct, I would rename this property `stackSetRegion`, and change these docs.

Reply:
I don't agree here, actually (for the `CloudFormationStackInstancesAction`).

The difference between this Action, and literally every other Action in this library so far, is that it manipulates multiple resources (the Stacks), and also does it in multiple regions. So, having a `region` parameter here is super confusing (it definitely confused me, and I wouldn't exactly call myself a novice in this module 😉).

Given that, I think renaming this property to `stackSetRegion` is the correct modeling (of course, pass it as `region` when calling `super()` in your constructor).